llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp

```diff
@@ -2538,6 +2538,70 @@ Value *LibCallSimplifier::optimizeLog(CallInst *Log, IRBuilderBase &B) {
   return Ret;
 }
 
+// sqrt(exp(X)) -> exp(X * 0.5)
+Value *LibCallSimplifier::mergeSqrtToExp(CallInst *CI, IRBuilderBase &B) {
+  if (!CI->isFast())
```

Comment:
I think you just need reassoc?

Reply:
I'm not sure whether just reassoc is enough. Don't we need afn (approximate calculations for functions) as well (or only afn)?

From langref for sqrt: `When specified with the fast-math-flag ‘afn’, the result may be approximated using a less accurate calculation`. https://llvm.org/docs/LangRef.html#llvm-sqrt-intrinsic
Or another example from clang users manual for `-fapprox-func`: `For example, a pow(x, 0.25) may be replaced with sqrt(sqrt(x))`. https://clang.llvm.org/docs/UsersManual.html#cmdoption-f-no-approx-func

Reply:
I did some experiments and this is net error reducing. This reassociation avoids some intermediate overflows to infinity and overall reduces worst case ulp. I think just reassoc is adequate

Reply:
@arsenm, thanks for the research. I've changed the code to check only reassoc flag.